File: Algorithm_Practice/c++/2023년 하반기/광물_캐기.java
import java.io.*;

//백트래킹, 재귀
//다이아 곡괭이 사용하는 경우, 철 곡괭이 사용하는 경우, 돌 곡괭이 사용하는 경우 모두 탐색
//주의! 인자값 받는 변수를 새로 선언해서 중간에 값이 바뀌어서 다음 재귀로 들어가지 않도록 주의하기
class Solution {
    public int countTired(String tool, String mineral){
        
        if(tool.equals("iron")){
            if(mineral.equals("diamond")){
                return 5;
            }
        }
        else if(tool.equals("stone")){
            if(mineral.equals("diamond")){
                return 25;
            }
            else if(mineral.equals("iron")){
                return 5;
            }
        }
        
        return 1;
    }
    
    public int answer = 25*50;
    
    public void backtracking(int diamondCnt, int ironCnt, int stoneCnt, int mineralIdx, String[] minerals, int tired){
        //System.out.println("diamond 곡괭이 "+diamondCnt+" iron 곡괭이 "+ironCnt+" 돌 곡괭이 "+stoneCnt+" "+mineralIdx+" 피로도 "+tired);
        
        //곡괭이를 다 썼을 때
        if(diamondCnt==0 && ironCnt==0 && stoneCnt==0){
            answer = Math.min(answer, tired);
            return;
        }
        
        //광물을 다 캤을 때
        if(mineralIdx>=minerals.length){
            answer = Math.min(answer, tired);
            return;
        }
        
        if(diamondCnt>0){
            //System.out.println("mineralIdx "+mineralIdx);
            int newTired = tired;
            int newMineralIdx = mineralIdx;
            for(int i=0; i<5; i++){
                if(newMineralIdx+i>=minerals.length){
                    //System.out.println("answer "+answer);
                    answer = Math.min(answer, newTired);
                    break;
                }
                newTired += countTired("diamond", minerals[newMineralIdx+i]);
            }
            newMineralIdx = mineralIdx + 5;
            backtracking(diamondCnt-1, ironCnt, stoneCnt, newMineralIdx, minerals, newTired);
        }
        if(ironCnt>0){
            //System.out.println("mineralIdx "+mineralIdx);
            int newTired = tired;
            int newMineralIdx = mineralIdx;
            for(int i=0; i<5; i++){
                if(newMineralIdx+i>=minerals.length){
                    answer = Math.min(answer, newTired);
                    break;
                }
                newTired += countTired("iron", minerals[newMineralIdx+i]);
            }
            newMineralIdx = mineralIdx + 5;
            backtracking(diamondCnt, ironCnt-1, stoneCnt, newMineralIdx, minerals, newTired);
        }
        if(stoneCnt>0){
            //System.out.println("mineralIdx "+mineralIdx);
            int newTired = tired;
            int newMineralIdx = mineralIdx;
            for(int i=0; i<5; i++){
                if(newMineralIdx+i>=minerals.length){
                    answer = Math.min(answer, newTired);
                    break;
                }
                newTired += countTired("stone", minerals[newMineralIdx+i]);
            }
            newMineralIdx = mineralIdx + 5;
            backtracking(diamondCnt, ironCnt, stoneCnt-1, newMineralIdx, minerals, newTired);
        }
    }
    
    public int solution(int[] picks, String[] minerals) {
        
        backtracking(picks[0], picks[1], picks[2], 0, minerals, 0);
        
        return answer;
    }
}

File: Algorithm_Practice/c++/2023년 하반기/광물_캐기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

//곡괭이를 사용할 수 있는 모든 경우에 피로도 계산
//next_permutation
int countTired(string tool, string mine){
    if(tool==mine){
        return 1;
    }
    
    if(tool=="diamond"){
        return 1;
    }
    else if(tool=="iron"){
        if(mine=="diamond"){
            return 5;
        }
        return 1;
    }
    else{
        if(mine=="diamond"){
            return 25;
        }
        else{
            return 5;
        }
    }
    
    return 0;
}

//곡괭이 사용할 수 있는 모든 경우의 수 구하기 - next_permutation
int solution(vector<int> picks, vector<string> minerals) {
    int answer = 25*50;
    vector<string> tools;
    
    for(int i=0; i<picks[0]; i++){
        tools.push_back("diamond");
    }
    for(int i=0; i<picks[1]; i++){
        tools.push_back("iron");
    }
    for(int i=0; i<picks[2]; i++){
        tools.push_back("stone");
    }
    
    //곡괭이 사용하는 모든 경우의 수
    do{
        int tired = 0;
        int first = 0;
        for(int i=0; i<tools.size(); i++){
            bool flag = true;
            for(int j=0; j<5; j++){
                if(first+j<minerals.size())
                    tired += countTired(tools[i], minerals[first+j]);
                else{
                    flag = false;
                    break;
                }
            }
            
            if(!flag){
                break;
            }
            
            first+=5;
        }
        answer = min(answer, tired);
    }
    while(next_permutation(tools.begin(), tools.end()));
    
    return answer;
}